rollout/trafficrouting.go

```diff
@@ -122,9 +112,33 @@ func (c *rolloutContext) reconcileTrafficRouting() error {
 			// last setWeight step, which is set by GetCurrentSetWeight.
 			desiredWeight = replicasetutil.GetCurrentSetWeight(c.rollout)
 		}
+
+		// Checks for experiment step
+		// If current experiment exists, then create WeightDestinations for each experiment template
+		exStep := replicasetutil.GetCurrentExperimentStep(c.rollout)
+		if exStep != nil && c.currentEx != nil {
+			getTemplateWeight := func(name string) *int32 {
+				for _, tmpl := range exStep.Templates {
+					if tmpl.Name == name {
+						return tmpl.Weight
+					}
+				}
+				return nil
+			}
+			for _, templateStatus := range c.currentEx.Status.TemplateStatuses {
+				templateWeight := getTemplateWeight(templateStatus.Name)
+				if templateStatus.ServiceName != "" || templateStatus.PodTemplateHash != "" {
+					weightDestinations = append(weightDestinations, trafficrouting.WeightDestination{
+						ServiceName:     templateStatus.ServiceName,
+						PodTemplateHash: templateStatus.PodTemplateHash,
+						Weight:          *templateWeight,
+					})
+				}
+			}
+		}
```

Comment:
What prevents us from prematurely setting the weight to the templates before the ReplicaSet is available?

Reply:
I added L130 because the trafficRouting providers were setting backends/annotations for the weightDestinations with an empty service.

Reply:
Are you implying that `.ServiceName` and `.PodTemplateHash` fields are only set when the Pods of the template are fully available?

Shouldn't we be checking the Running status of the experiment somewhere?

Reply:
Also, if experiment fails and the replicasets scale down. Do we still split traffic to the scaled down experiments?

Reply:
`.ServiceName` and `.PodTemplateHash` are set when the ReplicaSet is initialized with zero replicas, not when the Pods are fully available. When the ReplicaSet scales down, the service is deleted and `.ServiceName` is set to an empty string, so L130 prevents traffic from being sent to scaled down replicas.

Reply:
Should we be checking the Running status of the experiment or each experiment template?

Reply:
I would say the entire Experiment should be `Running` in order to consider splitting weights to its templates. Conversely, when the Experiment exits out of `Running`, then we should **_not_** split traffic to it. The Running status of the Experiment should be sufficient to know it is safe to target. I think checking against ServiceName and .PodTemplateHash are a round-about way of checking if it is safe, and it also doesn't cover the case when Experiment is still Pending.